### PR TITLE
Update: Add support for React (#217, #231)

### PR DESCRIPTION
**Pull request by taylortom**
#231, #217

Duplicate of https://github.com/adapt-security/adapt-authoring-ui/pull/232, for the UI update branch.

### Update
- use [rollup ](https://rollupjs.org/) as the bundler rather than [requirejs](https://requirejs.org)
- be able to fully utilise ES6 and beyond
- be backwards compatible with existing code base
- include [React](https://react.dev/) support
- compile sources in situ rather than copying them
- watch sources and compile on change
- no longer need to restart backend stack when working on UI
- cache bundler output for faster builds

Much of this work has been inspired by the [Adapt framework](https://github.com/adaptlearning/adapt_framework).

### Instructions
- checkout [issue/231](https://github.com/adapt-security/adapt-authoring-ui/tree/issue/231)
- run `npm start`
- wait for UI successfully built notice
- edit `js/jsx/hbs*/less` and observe console
- reload browser to see changes

*core `index.hbs`/`loading.hbs` are not pre-compiled and will be cached by the express templating engine; so in the (unlikely) event these are edited the process may need to be restarted

### Notes
- `npm start -- --disable-cache` replaces `npm start -- --rebuild-ui`
- [Example React conversion of appHeader](https://github.com/adapt-security/adapt-authoring-ui/commit/ede3c3aa34b61154c5aeabf0514218ff486b9085)

### CODE REVIEW

File: lib/JavaScriptTask.js [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+
+      const umdImport = () => {
+        return options.umdImports.map(filename => {
+          let code = fs.readFileSync(filename).toString();
+          code = code
+            .replace('require("object-assign")', 'Object.assign')
+            .replace('define.amd', 'define.noop');
+          return code;
+        }).join('\n');
+      };
```
Comment (oliverfoster) [inline lib/JavaScriptTask.js:418]: This and 56 umdImports: [] and 429 intro: umdImport() should not be necessary.

File: lib/JavaScriptTask.js [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+      externalMap:{},
+      map:{
+        "core": "adapt-authoring-ui/app/core",
+        "modules": "adapt-authoring-ui/app/modules"
+      },
+      umdImports: []
+    }
+
+    console.log(`Cache disabled (--disable-cache): ${this.isDisableCache}`);
+    console.log(`Strict mode (config.json:build.strictMode): ${this.buildConfig.strictMode}`);
```
Comment (oliverfoster) [inline lib/JavaScriptTask.js:60]: You can remove all the strictMode stuff and just default to strictMode. It was for ie11 which the AAT doesn't support.

**Review by oliverfoster — COMMENTED** [review]
How's the sourcemapping coming out?

**Comment by chris-steele** [comment]
Just seen this PR. thanks for the suggestion @oliverfoster I'll sort those out (and source mapping is working just fine thankfully!). @taylortom FYI new-build-process branches from RC2, whereas issue/231 was branched off master. For no-sidebar is new-build-process the best choice?

**Comment by chris-steele** [comment]
Please see https://github.com/adapt-security/adapt-authoring-ui/pull/243

**Pull request closed without merging**